### ISSUE TITLE
Report warning when dictionary or list is modified during iteration

### ISSUE THREAD

**Issue by yugr** [PyCQA/pylint#2471]
### Is your feature request related to a problem? Please describe

A frequent bug which I see people running into is modifying list during iteration. Example bug from [this](https://stackoverflow.com/questions/2896752/removing-item-from-list-during-iteration-whats-wrong-with-this-idiom) stackoverflow question:
```
letters=['a','b','c','d','e','f','g','h','i','j','k','l']
for i in letters:
    letters.remove(i)
print(letters)  # Prints ['b', 'd', 'f', 'h', 'j', 'l']
```

It would be great it if Pylint could detect this.

### Describe the solution you'd like

Pylint should emit a warning when program modifies list or dictionary object while iterating over it.

### Additional context

I've attached an example patch. For the following code
```
x = []
for i in x:
    x = [1, 2, 3]
```
it reports
```
tmp2.py:3:4: E0130: Iterated object 'x' is modified inside the loop (changed-iteratee)
```
The patch obviously needs to be fixed/extended, I'd be happy to work on that if the project is interested in this feature.

[0001-Initial-commit.patch.txt](https://github.com/PyCQA/pylint/files/2347428/0001-Initial-commit.patch.txt)

**Comment by rowlap** [comment]
Disagree with this not being useful - came to file this as a feature request, but found this issue. It would have saved several hours debugging.

**Comment by esmail** [comment]
Concretely, this would help prevent against e.g. `RuntimeError: dictionary keys changed during iteration`

**Comment by yugr** [comment]
@esmail my understanding of Pylint's development policy is that it should be a simple linter, capable of detecting syntax violations and relatively simple bugs. For realistic static analysis of Python code people should use different tools.

**Comment by DanielNoord** [comment]
Wasn't this added with https://github.com/PyCQA/pylint/pull/5628?

**Comment by Pierre-Sassoulas** [comment]
@DanielNoord is right, the code given as an example is raising a warning since 2.13.0:

> a.py:3:4: W4701: Iterated list 'letters' is being modified inside for loop body, consider iterating through a copy of it instead. (modified-iterating-list)


I'm going to close, please open another issue for false negatives or false positive related to the ``modified-iterating-list`` messages it something is missing.

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Did you also check if we special-case `delete`?

**Comment by Pierre-Sassoulas** [comment]
Yes, it does not work https://github.com/PyCQA/pylint/issues/6648 😄 

```python
x = []
for i in x:
    x = [1, 2, 3]
```    
Does not raise either.

**Comment by Pierre-Sassoulas** [comment]
Created a branch with the initial patch : https://github.com/PyCQA/pylint/tree/first-patch-2471

**Comment by jacobtylerwalls** [comment]
https://github.com/PyCQA/pylint/issues/2471#issuecomment-1131321534 is an interesting case, as I don't know whether it can lead to unexpected results at runtime. Bad style for sure, though.

**Comment by DanielNoord** [comment]
That sounds more like `redefined-outer-name` or whatever that message is called atm. 😄

**Comment by jacobtylerwalls** [comment]
Well, and the norm for that msg (currently) is to not emit when overwriting names in the same namespace[0], so that's maybe a no-op. I'm tempted to re-close this.

[0] -- it was the `redefined-loop-name` extension that would go further, and check for overwriting the `i` of the loop variable, not the `my_list` being iterated.

**Comment by jacobtylerwalls** [comment]
Thanks all -- if you find new cases to discuss, please open new issues.
